crates/worker/src/partition/shuffle.rs

```diff
@@ -276,148 +272,174 @@ mod state_machine {
         ),
     >;
 
+    async fn get_next_message<OutboxReader: super::OutboxReader>(
+        mut outbox_reader: OutboxReader,
+        next_sequence_number: MessageIndex,
+    ) -> (
+        Result<Option<(MessageIndex, OutboxMessage)>, OutboxReaderError>,
+        OutboxReader,
+    ) {
+        let result = outbox_reader.get_next_message(next_sequence_number).await;
+        (result, outbox_reader)
+    }
+
     #[pin_project(project = StateProj)]
-    enum State<SendFuture> {
+    enum State {
         Idle,
         ReadingOutbox,
-        Sending(#[pin] SendFuture),
+        Ingesting { ingest: IngestFuture, sn: u64 },
     }
 
     #[pin_project]
-    pub(super) struct StateMachine<'a, OutboxReader, SendOp, SendFuture> {
+    pub struct ShuffleStream<T, R> {
         metadata: ShuffleMetadata,
-        current_sequence_number: MessageIndex,
-        outbox_reader: Option<OutboxReader>,
-        read_future: ReadFuture<OutboxReader>,
-        send_operation: SendOp,
-        hint_rx: &'a mut async_channel::Receiver<NewOutboxMessage>,
+        ingestion: IngestionClient<T, Envelope>,
         #[pin]
-        state: State<SendFuture>,
-    }
-
-    async fn get_next_message<OutboxReader: shuffle::OutboxReader>(
-        mut outbox_reader: OutboxReader,
-        sequence_number: MessageIndex,
-    ) -> (
-        Result<Option<(MessageIndex, OutboxMessage)>, OutboxReaderError>,
-        OutboxReader,
-    ) {
-        let result = outbox_reader.get_next_message(sequence_number).await;
-        (result, outbox_reader)
+        hint_rx: async_channel::Receiver<NewOutboxMessage>,
+        reader: Option<R>,
+        read_fut: ReadFuture<R>,
+        inflight: VecDeque<RecordCommit<MessageIndex>>,
+        next_sequence_number: MessageIndex,
+        #[pin]
+        state: State,
```

Comment:
We probably can get rid of the whole pin business if we no longer poll the stream in the tokio::select! arm in the `run` method. Then we could simply have an `async` method which implements the logic. This could be simpler to implement than implementing the poll methods.